cdap-watchdog/src/main/java/co/cask/cdap/logging/write/LogCleanup.java

```diff
@@ -57,28 +64,45 @@ public void run() {
+      final Map<String, NamespaceId> namespacedLogBaseDirMap = new HashMap<>();
       fileMetaDataManager.cleanMetaData(tillTime,
                                         new FileMetaDataManager.DeleteCallback() {
                                           @Override
-                                          public void handle(Location location, String namespacedLogBaseDir) {
+                                          public void handle(NamespaceId namespaceId, final Location location,
+                                                             String namespacedLogBaseDir) {
                                             try {
-                                              if (location.exists()) {
-                                                LOG.info("Deleting log file {}", location);
-                                                location.delete();
-                                              }
+                                              impersonator.doAs(namespaceId, new Callable<Void>() {
+                                                @Override
+                                                public Void call() throws Exception {
+                                                  if (location.exists()) {
+                                                    LOG.info("Deleting log file {}", location);
+                                                    location.delete();
+                                                  }
+                                                  return null;
+                                                }
+                                              });
                                               parentDirs.put(namespacedLogBaseDir, getParent(location));
-                                            } catch (IOException e) {
+                                              namespacedLogBaseDirMap.put(namespacedLogBaseDir, namespaceId);
+                                            } catch (Exception e) {
                                               LOG.error("Got exception when deleting path {}", location, e);
                                               throw Throwables.propagate(e);
                                             }
                                           }
                                         });
       // Delete any empty parent dirs
-      for (String namespacedLogBaseDir : parentDirs.keySet()) {
-        Set<Location> locations = parentDirs.get(namespacedLogBaseDir);
-        for (Location location : locations) {
-          deleteEmptyDir(namespacedLogBaseDir, location);
-        }
+      for (final String namespacedLogBaseDir : parentDirs.keySet()) {
+        // this ensures that we only do doAs which will make an RPC call only once for a namespace
+        NamespaceId namespaceId = namespacedLogBaseDirMap.get(namespacedLogBaseDir);
+        impersonator.doAs(namespaceId, new Callable<Void>() {
```

Comment:
we don't need to wrap these operations within a `doAs`. Just the `Location` object has to be created within a `doAs`.